File: src/rust/iced-x86-py/tests/encoder_test.py
# SPDX-License-Identifier: MIT
# Copyright (C) 2018-present iced project and contributors

import pytest
from iced_x86 import *

@pytest.mark.parametrize("bitness", [16, 32, 64, 0, 15, 128])
def test_invalid_bitness(bitness):
	if bitness == 16 or bitness == 32 or bitness == 64:
		Encoder(bitness)
	else:
		with pytest.raises(ValueError):
			Encoder(bitness)

@pytest.mark.parametrize("capacity", [0, 1, 0x1234])
def test_capacity_arg(capacity):
	Encoder(64, capacity)

@pytest.mark.parametrize("bitness", [16, 32, 64])
def test_bitness(bitness):
	encoder = Encoder(bitness)
	assert encoder.bitness == bitness

@pytest.mark.parametrize("bitness", [16, 32, 64])
def test_options(bitness):
	encoder = Encoder(bitness)
	assert not encoder.prevent_vex2
	assert encoder.vex_wig == 0
	assert encoder.vex_lig == 0
	assert encoder.evex_wig == 0
	assert encoder.evex_lig == 0

	encoder.prevent_vex2 = True
	encoder.vex_wig = 1
	encoder.vex_lig = 1
	encoder.evex_wig = 1
	encoder.evex_lig = 2

	assert encoder.prevent_vex2
	assert encoder.vex_wig == 1
	assert encoder.vex_lig == 1
	assert encoder.evex_wig == 1
	assert encoder.evex_lig == 2

@pytest.mark.parametrize("bitness, data", [
	(16, b"\x03\xCE\x90\xF3\x90"),
	(32, b"\x03\xCE\x90\xF3\x90"),
	(64, b"\x48\x09\xCE\x90\xF3\x90"),
])
def test_encode(bitness, data):
	encoder = Encoder(bitness)
	decoder = Decoder(bitness, data, ip=0x1234_5678_9ABC_DEF0)
	rip = decoder.ip
	for instr in decoder:
		instr_len = encoder.encode(instr, rip)
		assert instr.len == instr_len
		rip += instr_len
	encoded_data = encoder.take_buffer()
	assert type(encoded_data) == bytes
	assert data == encoded_data

def test_encode_rip_63_set():
	decoder = Decoder(64, b"\x48\x09\xCE\x90\xF3\x90")
	instr = decoder.decode()
	encoder = Encoder(64)
	assert encoder.encode(instr, 0x1234_5678_9ABC_DEF0) == instr.len
	assert encoder.encode(instr, 0xFEDC_BA98_7654_3210) == instr.len

def test_write_u8():
	decoder = Decoder(64, b"\x48\x09\xCE\x90\xF3\x90")
	instr = decoder.decode()
	encoder = Encoder(64)
	encoder.write_u8(0x12)
	encoder.write_u8(0x34)
	assert encoder.encode(instr, 0) == 3
	encoder.write_u8(0x56)
	encoder.write_u8(0x78)
	encoder.write_u8(0x9A)
	encoded_data = encoder.take_buffer()
	assert encoded_data == b"\x12\x34\x48\x09\xCE\x56\x78\x9A"

def test_encode_invalid_instruction():
	# Can't encode INVALID
	instr = Instruction()
	encoder = Encoder(64)
	with pytest.raises(ValueError):
		encoder.encode(instr, instr.ip)

	# Jcc SHORT with a target too far away
	decoder = Decoder(64, b"\x72\x00", ip=0x1234_5678_9ABC_DEF0)
	decoder.decode_out(instr)
	encoder.encode(instr, instr.ip)
	with pytest.raises(ValueError):
		encoder.encode(instr, 0xFEDC_BA98_7654_3210)

def test_offsets():
	decoder = Decoder(64, b"\x90\x83\xB3\x34\x12\x5A\xA5\x5A")
	encoder = Encoder(64)

	instr = decoder.decode()
	assert encoder.encode(instr, instr.ip) == 1
	co = encoder.get_constant_offsets()
	assert co.displacement_offset == 0
	assert co.displacement_size == 0
	assert co.immediate_offset == 0
	assert co.immediate_size == 0
	assert co.immediate_offset2 == 0
	assert co.immediate_size2 == 0
	assert not co.has_displacement
	assert not co.has_immediate
	assert not co.has_immediate2

	instr = decoder.decode()
	assert encoder.encode(instr, instr.ip) == 7
	co = encoder.get_constant_offsets()
	assert co.displacement_offset == 2
	assert co.displacement_size == 4
	assert co.immediate_offset == 6
	assert co.immediate_size == 1
	assert co.immediate_offset2 == 0
	assert co.immediate_size2 == 0
	assert co.has_displacement
	assert co.has_immediate
	assert not co.has_immediate2

File: src/rust/iced-x86-js/src/tests/encoder.test.js
// SPDX-License-Identifier: MIT
// Copyright (C) 2018-present iced project and contributors

const { Decoder, DecoderOptions, Encoder, getIcedFeatures } = require("iced-x86");

test("Creating an Encoder with an invalid bitness throws", () => {
	expect(() => new Encoder(63)).toThrow();
	expect(() => Encoder.withCapacity(63, 64)).toThrow();
});

test("Encoder options", () => {
	const encoder = new Encoder(64);

	expect(encoder.preventVEX2).toBe(false);
	encoder.preventVEX2 = true;
	expect(encoder.preventVEX2).toBe(true);

	expect(encoder.VEX_WIG).toBe(0);
	encoder.VEX_WIG = 1;
	expect(encoder.VEX_WIG).toBe(1);

	expect(encoder.VEX_LIG).toBe(0);
	encoder.VEX_LIG = 1;
	expect(encoder.VEX_LIG).toBe(1);

	expect(encoder.EVEX_WIG).toBe(0);
	encoder.EVEX_WIG = 1;
	expect(encoder.EVEX_WIG).toBe(1);

	expect(encoder.EVEX_LIG).toBe(0);
	encoder.EVEX_LIG = 1;
	expect(encoder.EVEX_LIG).toBe(1);
	encoder.EVEX_LIG = 2;
	expect(encoder.EVEX_LIG).toBe(2);
	encoder.EVEX_LIG = 3;
	expect(encoder.EVEX_LIG).toBe(3);

	// Check if MVEX support
	if ((getIcedFeatures() & 0x10) != 0) {
		expect(encoder.MVEX_WIG).toBe(0);
		encoder.MVEX_WIG = 1;
		expect(encoder.MVEX_WIG).toBe(1);
	}

	encoder.free();
});

test("16-bit encoder", () => {
	const bytes = new Uint8Array([0x75, 0x02]);
	const decoder = new Decoder(16, bytes, DecoderOptions.None);
	decoder.ip = 0x8123n;
	const instr = decoder.decode();
	const encoder = new Encoder(16);

	expect(encoder.bitness).toBe(16);

	encoder.writeU8(0xCC);
	expect(encoder.encode(instr, 0x8120n)).toBe(2);
	encoder.writeU8(0x90);

	const buffer = encoder.takeBuffer();
	expect(encoder.takeBuffer().length).toBe(0);
	encoder.setBuffer(new Uint8Array([0x11, 0x22, 0x33]));
	expect(encoder.takeBuffer()).toStrictEqual(new Uint8Array([0x11, 0x22, 0x33]));
	expect(buffer).toStrictEqual(new Uint8Array([0xCC, 0x75, 0x05, 0x90]));

	decoder.free();
	instr.free();
	encoder.free();
});

test("32-bit encoder", () => {
	const bytes = new Uint8Array([0x75, 0x02]);
	const decoder = new Decoder(32, bytes, DecoderOptions.None);
	decoder.ip = 0x81234567n;
	const instr = decoder.decode();
	const encoder = new Encoder(32);

	expect(encoder.bitness).toBe(32);

	encoder.writeU8(0x90);
	expect(encoder.encode(instr, 0x81234563n)).toBe(2);
	encoder.writeU8(0xCC);

	const buffer = encoder.takeBuffer();
	expect(encoder.takeBuffer().length).toBe(0);
	encoder.setBuffer(new Uint8Array([0x11, 0x22, 0x33]));
	expect(encoder.takeBuffer()).toStrictEqual(new Uint8Array([0x11, 0x22, 0x33]));
	expect(buffer).toStrictEqual(new Uint8Array([0x90, 0x75, 0x06, 0xCC]));

	decoder.free();
	instr.free();
	encoder.free();
});

test("64-bit encoder", () => {
	const bytes = new Uint8Array([0x75, 0x02]);
	const decoder = new Decoder(64, bytes, DecoderOptions.None);
	decoder.ip = 0x81234567ABCDEF01n;
	const instr = decoder.decode();
	const encoder = new Encoder(64);

	expect(encoder.bitness).toBe(64);

	encoder.writeU8(0xCD);
	expect(encoder.encode(instr, 0x81234567ABCDEEFFn)).toBe(2);
	encoder.writeU8(0x91);

	const buffer = encoder.takeBuffer();
	expect(encoder.takeBuffer().length).toBe(0);
	encoder.setBuffer(new Uint8Array([0x11, 0x22, 0x33]));
	expect(encoder.takeBuffer()).toStrictEqual(new Uint8Array([0x11, 0x22, 0x33]));
	expect(buffer).toStrictEqual(new Uint8Array([0xCD, 0x75, 0x04, 0x91]));

	decoder.free();
	instr.free();
	encoder.free();
});

test("Encoder constant offsets", () => {
	const bytes = new Uint8Array([0x90, 0x83, 0xB3, 0x34, 0x12, 0x5A, 0xA5, 0x5A]);
	const decoder = new Decoder(64, bytes, DecoderOptions.None);
	const instr1 = decoder.decode();
	const instr2 = decoder.decode();
	const encoder = Encoder.withCapacity(64, 100);

	encoder.encode(instr1, 0n);
	const co1 = encoder.getConstantOffsets();
	expect(co1.hasDisplacement).toBe(false);
	expect(co1.displacementOffset).toBe(0);
	expect(co1.displacementSize).toBe(0);
	expect(co1.hasImmediate).toBe(false);
	expect(co1.immediateOffset).toBe(0);
	expect(co1.immediateSize).toBe(0);
	expect(co1.hasImmediate2).toBe(false);
	expect(co1.immediateOffset2).toBe(0);
	expect(co1.immediateSize2).toBe(0);

	encoder.encode(instr2, 0n);
	const co2 = encoder.getConstantOffsets();
	expect(co2.hasDisplacement).toBe(true);
	expect(co2.displacementOffset).toBe(2);
	expect(co2.displacementSize).toBe(4);
	expect(co2.hasImmediate).toBe(true);
	expect(co2.immediateOffset).toBe(6);
	expect(co2.immediateSize).toBe(1);
	expect(co2.hasImmediate2).toBe(false);
	expect(co2.immediateOffset2).toBe(0);
	expect(co2.immediateSize2).toBe(0);

	decoder.free();
	encoder.free();
	instr1.free();
	instr2.free();
	co1.free();
	co2.free();
});
